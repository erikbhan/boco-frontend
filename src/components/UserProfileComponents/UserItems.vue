<template>
  <div id = "headline" class = "text-xl md:text-2xl text-gray-600 font-medium w-full">
      Dine gjenstander
  </div>
  <!-- Search field -->
    <div class="relative" id="searchComponent">
      <span class="absolute inset-y-0 left-0 flex items-center pl-3">
        <svg class="w-5 h-5 text-gray-400" viewBox="0 0 24 24" fill="none">
          <path
            d="M21 21L15 15M17 10C17 13.866 13.866 17 10 17C6.13401 17 3 13.866 3 10C3 6.13401 6.13401 3 10 3C13.866 3 17 6.13401 17 10Z"
            stroke="currentColor"
            stroke-width="2"
            stroke-linecap="round"
            stroke-linejoin="round"
          ></path>
        </svg>
      </span>

      <input
        type="text"
        id="searchInput"
        class="w-full py-3 pl-10 pr-4 text-gray-700 bg-white border rounded-md dark:bg-gray-800 dark:text-gray-300 dark:border-gray-600 focus:border-primary-medium dark:focus:border-primary-medium focus:outline-none focus:ring"
        placeholder="Search"
        v-model="search"
        @change="searchWritten"
      />
    </div>

    <div class="absolute inset-x-0 px-5 py-3">
      <!-- ItemCards -->
      <div class="flex items-center justify-center w-screen">
        <!-- Shows items based on pagination -->
        <div
          class="grid grid-flow-row-dense grid-cols-2 md:grid-cols-4 lg:grid-cols-5 w-full"
          v-if="showItems"
        >
          <ItemCard
            v-for="item in visibleItems"
            :key="item"
            :item="item"
          />
        </div>

        <!-- Shows items based on search field input -->
        <div
          class="grid grid-flow-row-dense grid-cols-2 md:grid-cols-4 lg:grid-cols-5 w-full place-items-center"
          v-if="showSearchedItems"
        >
          <ItemCard
            v-for="item in searchedItems"
            :key="item"
            :item="item"
          />
        </div>
      </div>
      <!-- pagination -->
      <div class="flex justify-center" v-if="showItems">
        <PaginationTemplate
          v-bind:items="items"
          v-on:page:update="updatePage"
          v-bind:currentPage="currentPage"
          v-bind:pageSize="pageSize"
          class="mt-10"
        />
      </div>
    </div>
</template>
<script>
import { GetUserListings, getItemPictures } from "@/utils/apiutil";
import ItemCard from "@/components/ItemComponents/ItemCard.vue";
import PaginationTemplate from "@/components/BaseComponents/PaginationTemplate";

export default {
  name: "UserItems",
  components: {
    ItemCard,
    PaginationTemplate
  },
  data() {
    return {
      items: [],
      item: {
        listingID: 0,
        img: "",
        address: "",
        title: "",
        pricePerDay: 0,
      },
      showItems: true,
      showSearchedItems: false,
        search: "",
      //Variables connected to pagination
      currentPage: 0,
      pageSize: 12,
      visibleItems: [],
    };
  },
  computed: {
    searchedItems() {
      let filteredItems = [];

      filteredItems = this.items.filter(
        (p) =>
          p.title.toLowerCase().includes(this.search.toLowerCase()) ||
          p.address.toLowerCase().includes(this.search.toLowerCase()) ||
          p.pricePerDay === Number(this.search)
      );

      return filteredItems;
    },
  },
  methods: {
    getUserListingsFromAPI: async function () {
      this.items = await GetUserListings();
      for (var i = 0; i < this.items.length; i++) {
        let images = await getItemPictures(this.items[i].listingID);
        if (images.length > 0) {
          this.items[i].img = images[0].picture;
        }
      }
    },
      //Pagination
    updatePage(pageNumber) {
      this.currentPage = pageNumber;
      this.updateVisibleTodos();
    },
    updateVisibleTodos() {
      this.visibleItems = this.items.slice(
        this.currentPage * this.pageSize,
        this.currentPage * this.pageSize + this.pageSize
      );

      // if we have 0 visible items, go back a page
      if (this.visibleItems.length === 0 && this.currentPage > 0) {
        this.updatePage(this.currentPage - 1);
      }
    },
    searchWritten: function () {
      //This method triggers when search input field is changed
      if (this.search.length > 0) {
        this.showItems = false;
        this.showSearchedItems = true;
      } else {
        this.showItems = true;
        this.showSearchedItems = false;
      }
    },
  },
  async beforeMount() {
    await this.getUserListingsFromAPI();
    this.updateVisibleTodos();

  },
};
</script>

<style>
#headline {
  display: block;
  margin-top: 10px;
  margin-bottom: 10px;  
  
}
</style>