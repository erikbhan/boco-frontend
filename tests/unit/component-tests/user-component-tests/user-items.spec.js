import { mount } from "@vue/test-utils";
import UserItems from "@/components/UserProfileComponents/UserItems.vue";

describe("UserItems component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(UserItems, {
      data() {
        return {
          items: [
            {
              listingID: 1,
              img: "",
              address: "Veien",
              title: "Matboks",
              pricePerDay: 50,
              toggle: false,
            },
          ],
        };
      },
    });
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("is instantiated", () => {
    expect(wrapper.exists()).toBeTruthy();
  });

  it("Check headline", () => {
    expect(wrapper.find("#headline").text()).toMatch("Mine gjenstander");
  });
});
