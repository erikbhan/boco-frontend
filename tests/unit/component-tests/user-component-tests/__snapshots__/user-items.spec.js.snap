// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserItems component renders correctly 1`] = `
<div
  data-v-app=""
>
  
  <!-- Shows all the items a user has posted with search and pagination.
       Includes a dropdown menu for editing or deleting an item. -->
  <div
    class="text-xl md:text-2xl text-primary-light font-medium"
    id="headline"
  >
     Mine gjenstander 
  </div>
  <!-- Search field -->
  <div
    class="relative mx-4"
    id="searchComponent"
  >
    <span
      class="absolute inset-y-0 left-0 flex items-center pl-3"
    >
      <svg
        class="w-5 h-5 text-gray-400"
        fill="none"
        viewBox="0 0 24 24"
      >
        <path
          d="M21 21L15 15M17 10C17 13.866 13.866 17 10 17C6.13401 17 3 13.866 3 10C3 6.13401 6.13401 3 10 3C13.866 3 17 6.13401 17 10Z"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </span>
    <input
      class="w-full py-3 pl-10 pr-4 text-gray-700 bg-white border rounded-md dark:bg-gray-800 dark:text-gray-300 dark:border-gray-600 focus:border-primary-medium dark:focus:border-primary-medium focus:outline-none focus:ring"
      id="searchInput"
      placeholder="SÃ¸k"
      type="text"
    />
  </div>
  <div
    class="absolute inset-x-0"
  >
    <!-- ItemCards -->
    <div
      class="flex items-center justify-center w-screen"
    >
      <!-- Shows items based on pagination -->
      <div
        class="grid grid-flow-row-dense grid-cols-2 md:grid-cols-4 lg:grid-cols-5 w-full"
      >
        
        
        <!-- A waring asking the user if it is sure it wants to delete the item
             with options to go ahead with the deleting or to cancel the delete. -->
        
        <!-- Main modal -->
        <!--v-if-->
        
      </div>
      <!-- Shows items based on search field input -->
      <!--v-if-->
    </div>
    <!-- pagination -->
    <div
      class="flex justify-center"
    >
      <div
        class="mt-10"
      >
        <!--v-if-->
        <label
          class="mx-2 text-primary-light"
        >
          1 av 1
        </label>
        <!--v-if-->
      </div>
    </div>
  </div>
  
</div>
`;
